test/system/src/test/java/io/pravega/test/system/StreamCutsTest.java

```diff
@@ -0,0 +1,383 @@
+                                                                                             .map(CompletableFuture::join)
+                                                                                             .reduce((a, b) -> a + b).get();
+                log.debug("Read events by group {}: {}.", readerGroupId, readEvents);
+                assertEquals("Expected events read: ", combinationCutSize * CUT_SIZE, readEvents);
+                combinationCutSize++;
+            }
+        }
+    }
+
+    /**
+     * Test that all the stream slices represented by consecutive StreamCut pairs can be read correctly.
+     *
+     * @param manager Group manager for this scope.
+     * @param clientFactory Client factory to instantiate new readers.
+     * @param streamSlices StreamCuts lists to be combined and tested via bounded processing.
+     */
+    private void readSliceBySliceAndVerify(ReaderGroupManager manager, ClientFactory clientFactory, int parallelSegments,
+                                           List<Map<Stream, StreamCut>> streamSlices) {
+        int readEvents;
+        for (int i = 1; i < streamSlices.size(); i++) {
+            ReaderGroupConfig configBuilder = ReaderGroupConfig.builder().stream(Stream.of(SCOPE, STREAM_ONE))
+                                                                         .stream(Stream.of(SCOPE, STREAM_TWO))
+                                                                         .startingStreamCuts(streamSlices.get(i - 1))
+                                                                         .endingStreamCuts(streamSlices.get(i)).build();
+
+            // Create a new reader group per stream cut slice and read in parallel only events within the cut.
+            final String readerGroupId = READER_GROUP + String.valueOf(i) + "-" + System.nanoTime();
+            manager.createReaderGroup(readerGroupId, configBuilder);
+            readEvents = readEventFutures(clientFactory, readerGroupId, parallelSegments).stream()
+                                                                                         .map(CompletableFuture::join)
+                                                                                         .reduce((a, b) -> a + b).get();
+            log.debug("Read events by group {}: {}.", readerGroupId, readEvents);
+            assertEquals("Expected events read: ", CUT_SIZE, readEvents);
+        }
+    }
+
+    private <T extends Serializable> List<Map<Stream, StreamCut>> getStreamCutSlices(ClientFactory client, ReaderGroup readerGroup,
+                                                                                     int totalEvents) {
+        @Cleanup
+        EventStreamReader<T> reader = client.createReader("slicer", readerGroup.getGroupName(), new JavaSerializer<>(),
+                ReaderConfig.builder().build());
+        List<Map<Stream, StreamCut>> streamCuts = new ArrayList<>();
+        EventRead<T> event;
+        int validEvents = 0;
+
+        do {
```

Comment:
To be consistent with other parts of the code, do you want to use our functional way of looping:

https://github.com/pravega/pravega/blob/master/common/src/main/java/io/pravega/common/concurrent/Futures.java#L593

Reply:
The loop is now implemented with `Futures.loop()`.